README.md
# jQuery Notify - A Notifier Plugin

[![Build Status](https://img.shields.io/travis/wbotelhos/notify/master.svg)](https://travis-ci.org/wbotelhos/notify "Travis CI")
[![Dependency Status](https://david-dm.org/wbotelhos/notify.svg)](https://david-dm.org/wbotelhos/notify "Dependency Status")
[![Dev Dependency Status](https://david-dm.org/wbotelhos/notify/dev-status.svg)](https://david-dm.org/wbotelhos/notify#info=devDependencies "Dev Dependency Status")
[![Code Climate](https://codeclimate.com/github/wbotelhos/notify.png)](https://codeclimate.com/github/wbotelhos/notify "Code Climate")
[![Support jQuery Notify](http://img.shields.io/gittip/wbotelhos.svg)](https://www.gittip.com/wbotelhos "Git Tip")

## Version

```
@version  0.1.0
@since    2010.08.03
@author   Washington Botelho
@doc      wbotelhos.com/notify
```

## Dependencies

+ jquery.js

## Usage

```html
<link rel="stylesheet" href="jquery.notify.css">
<script src="jquery.notify.js"></script>
```

```js
$.notify('Hello World!');
```

## Usage with Icon

- jquery.notify.[eot|svg|ttf|woff]

```html
<link rel="stylesheet" href="jquery.notify.css">
<link rel="stylesheet" href="jquery.notify.fonts.css">
<script src="jquery.notify.js"></script>
```

```js
$.notify({ body: 'Hello World!', icon: 'notify-info' });
```

## Bower Install

```js
bower install notify
```

See [bower.io](http://bower.io) for more information.

## Options

```js
afterClose   : undefined     // Callback executed after notice closes.
afterOpen    : undefined     // Callback executed after notice opens.
beforeClose  : undefined     // Callback executed before notice closes.
beforeOpen   : undefined     // Callback executed before notice opens.
body         : undefined     // Body of notice.
click        : undefined     // Callback executed when notice is clicked.
closeClick   : true          // If notice will be closed on click.
destroy      : false         // Removes the wrapper when there is no more notice on it.
forever      : false         // If notice will stay on screen forever.
hideTime     : 1000          // Time spent to hide the notice.
icon         : undefined     // Icon (font) name for the notice.
id           : 'notify-item' // Identifier to be used as class on notice.
image        : undefined     // Image path to be used into notice.
max          : 5             // Number maximum of noticies on screen.
minimizeTime : 300           // Time spent to minimize the notice.
mouseout     : undefined     // Callback executed on mouse over the notice.
mouseover    : undefined     // Callback executed on mouse out the notice.
position     : undefined     // The place where notices will appears.
showTime     : 400           // Time spent to show the notice.
timeout      : 5000          // Time spent to initiates the other timeouts.
title        : undefined     // The notice title.
wrapper      : '#notify'     // Where notices will be appended.
```

## Functions

First, save the notice reference:

```js
var notice = $.notify('Hello World!');
```

Then you can use the functions:

```js
notice.remove();               // Removes the notice from screen.

notice.position('class-name'); // Changes the notices position via CSS class.
```

## Global Functions

```js
$.notify.clean('name') // Clean the queue with name equal you passed or all when no name is given.
```

## Contributors

[Check it out](http://github.com/wbotelhos/notify/graphs/contributors)

## Licence

[The MIT License](http://opensource.org/licenses/MIT)

## Love it!

Via [PayPal](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=X8HEP2878NDEG&item_name=jQuery%20Notify) or [Gittip](http://www.gittip.com/wbotelhos). Thanks! (:


changelog.md
# jQuery Notify - A Notifier Plugin - http://wbotelhos.com/notify

## 0.1.0

### News

- Bower supported;
- JS Hint validation;
- Package.json on it;
- Tested with Jasmine.

#### Options

`afterClose` to be executed after notice closes.
`afterOpen` to be executed after notice opens.
`beforeClose` to be executed before close notice. Return `false` will prevent it;
`beforeOpen` to be executed before notice opens. Return `false` will prevent it;
`body` to choose the notice body;
`closeClick` to be possible close the notice with click;
`forever` to avoid the notice disappear;
`hideTime` to choose time in milliseconds to hide the notice;
`icon` to choose some icon on the left of notice;
`id` to change the name used to identify each notice item;
`image` to choose some image on the left of notice;
`minimizeTime` to choose the time in milliseconds to minimize the notice;
`position` to choose where notices will appears.
`showTime` to choose time in milliseconds to show the notice;
`timeout` to choose the time in milliseconds before `hideTime` starts;
`title` to choose the title of the notice;
`wrapper` to choose, or create, the wrapper where notices will be appended.

#### Functions

`close` to close the current notice instance;
`position` to change the position of noticies via class name.

#### Global Functions

`$.notify.clean()` to clean the queues `queue`, `lives` or both.


spec/karma.conf.js
module.exports = function(config) {
  config.set({
    preprocessors : {
      '../lib/*.js'        : ['jshint'],
      './integration/*.js' : ['jshint'],
      './model/*.js'       : ['jshint']
    },
    browsers      : ['Firefox'],
    files         : ['../vendor/*.js', '../lib/*.css', '../lib/*.js', 'lib/*.js', 'model/options_spec.js', '**/*spec.js' ],
    frameworks    : ['jasmine'],
    logLevel      : config.LOG_ERROR,
    reporters     : ['dots'],
    singleRun     : true
    // ,
    // jshint        : {
    //   options: {
    //     indent: 4,
    //     browser: true
    //   },
    //   summary: true
    // }
  });
};


spec/lib/helper.js
function context(description, spec) {
  describe(description, spec);
}

beforeEach(function() {
  $.notify.lives = [];
  $.notify.queue = [];
});

afterEach(function() {
  $($.notify.defaults.wrapper).remove();
});


lib/jquery.notify.js
function Notify(options, params) {
  'use strict';

  if (typeof(options) === 'string') {
    params      = params || {};
    params.body = options;
    options     = params;
  }

  this.opt     = $.extend(true, {}, $.notify.defaults, options);
  this.wrapper = this._wrapper();
  this.hash    = this._hash();
}

Notify.prototype._afterOpen = function() {
  'use strict';

  this._execute('afterOpen');
};

Notify.prototype._bindClick = function() {
  'use strict';

  this.item.on('click.notify', this._onClick.bind(this));
};

Notify.prototype._bindMouseenter = function() {
  'use strict';

  this.item.on('mouseenter.notify', this._onMouseenter.bind(this));
};

Notify.prototype._bindMouseleave = function() {
  'use strict';

  this.item.on('mouseleave.notify', this._onMouseleave.bind(this));
};

Notify.prototype._body = function() {
  'use strict';

  return $('<p />', { html: this.opt.body });
};

Notify.prototype._build = function() {
  'use strict';

  var item = this._item();

  if (this.opt.icon) {
    item.addClass('notify-icon').append(this._icon());
  } else if (this.opt.image) {
    item.addClass('notify-image').append(this._image());
  }

  if (this.opt.title) {
    item.append(this._title());
  }

  return item.append(this._body());
};

Notify.prototype._checkWrapper = function() {
  'use strict';

  if (!this.wrapper.children().length) {
    this.wrapper.remove();
  }
};

Notify.prototype._close = function(evt) {
  'use strict';

  if (this._execute('beforeClose', [evt])) {
    this.item.remove();

    this._pipe();

    if (this.opt.destroy) {
      this._checkWrapper();
    }

    this._execute('afterClose');
  }
};

Notify.prototype._create = function() {
  'use strict';

  this.item = this._build();

  this._bindClick();
  this._bindMouseenter();

  if (!this.opt.forever) {
    this._scheduleClose();
    this._bindMouseleave();
  }

  this._position(this.opt.position);
  this._open();
};

Notify.prototype._execute = function(name, args) {
  'use strict';

  var execute;

  if ($.isFunction(this.opt[name])) {
    execute = this.opt[name].apply(this, args);
  }

  return execute || execute === undefined;
};

Notify.prototype._hash = function() {
  'use strict';

  return +new Date();
};

Notify.prototype._hide = function() {
  'use strict';

  this.item.animate({ opacity: 0 }, this.opt.hideTime, this._minimize.bind(this));
};

Notify.prototype._icon = function() {
  'use strict';

  return $('<i />', { 'class': this.opt.icon });
};

Notify.prototype._image = function() {
  'use strict';

  return $('<img />', { src: this.opt.image });
};

Notify.prototype._item = function() {
  'use strict';

  return $('<div />', { 'class': this.opt.id });
};

Notify.prototype._minimize = function() {
  'use strict';

  this.item.animate({ height: 0 }, this.opt.minimizeTime, this._close.bind(this));
};

Notify.prototype._onClick = function(evt) {
  'use strict';

  if (this._execute('click', [evt]) && this.opt.closeClick) {
    this._close();
  }
};

Notify.prototype._onMouseenter = function(evt) {
  'use strict';

  if (this._execute('mouseover', [evt])) {
    this._unscheduleClose();
  }
};

Notify.prototype._onMouseleave = function(evt) {
  'use strict';

  if (this._execute('mouseout', [evt])) {
    this._scheduleClose();
  }
};

Notify.prototype._open = function() {
  'use strict';

  if (this._execute('beforeOpen')) {
    this._show();
  }
};

Notify.prototype._pipe = function() {
  'use strict';

  $.notify.lives.shift();

  var piped = $.notify.queue.shift();

  if (piped) {
    $.notify.lives.push(piped);
    piped._create();
  }
};

Notify.prototype._push = function() {
  'use strict';

  var ok = $.notify.lives.length < this.opt.max;

  if (ok) {
    $.notify.lives.push(this);
  } else {
    $.notify.queue.push(this);
  }

  return ok;
};

Notify.prototype._scheduleClose = function() {
  'use strict';

  this[this._timeoutKey()] = setTimeout(this._hide.bind(this), this.opt.timeout);
};

Notify.prototype._show = function() {
  'use strict';

  var that = this;

  that.item
  .appendTo(that.wrapper)
  .fadeIn(that.opt.showTime, this._afterOpen.bind(this));
};

Notify.prototype._timeoutKey = function() {
  'use strict';

  return 'notify_' + this.hash;
};

Notify.prototype._title = function() {
  'use strict';

  return $('<strong />', { html: this.opt.title });
};

Notify.prototype._position = function(position) {
  'use strict';

  if (position) {
    this.wrapper[0].className = position;
  }
};

Notify.prototype._unscheduleClose = function() { // TODO: CHAge to unschedule
  'use strict';

  clearTimeout(this[this._timeoutKey()]);

  this.item.stop().css({ opacity: '', height: '' });
};

Notify.prototype._wrapper = function() {
  'use strict';

  var wrapper = $(this.opt.wrapper);

  if (!wrapper.length) {
    return $('<div />', { id: this.opt.wrapper.slice(1) }).appendTo('body');
  }

  return wrapper;
};

Notify.prototype.close = function() {
  'use strict';

  this._close();
};

Notify.prototype.position = function(position) {
  'use strict';

  this._position(position);
};

(function($) {
  'use strict';

  $.notify = function(options, params) {
    var notify = new Notify(options, params);

    if (notify._push()) {
      notify._create();
    }

    return notify;
  };

  $.notify.clean = function(name) {
    if (!name || name === 'queue') {
      $.notify.queue = [];
    }

    if (!name || name === 'lives') {
      $.notify.lives = [];
    }
  };

  $.notify.queue = [];
  $.notify.lives = [];

  $.notify.defaults = {
    afterClose   : undefined,
    afterOpen    : undefined,
    beforeClose  : undefined,
    beforeOpen   : undefined,
    body         : undefined,
    click        : undefined,
    closeClick   : true,
    destroy      : false,
    forever      : false,
    hideTime     : 1000,
    icon         : undefined,
    id           : 'notify-item',
    image        : undefined,
    max          : 5,
    minimizeTime : 300,
    mouseout     : undefined,
    mouseover    : undefined,
    position     : undefined,
    showTime     : 400,
    timeout      : 5000,
    title        : undefined,
    wrapper      : '#notify'
  };
})(jQuery);
